vlm/src/main/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSource.scala

```diff
@@ -57,7 +57,11 @@ case class GeoTiffRasterSource(
       // TODO: shouldn't GridExtent give me Extent for types other than N ?
       Raster(tile, gridExtent.extentFor(gb.toGridType[Long], clamp = false))
     }
-    if (it.hasNext) Some(convertRaster(it.next)) else None
+
+    // We want to use this tiff in different `RasterSource`s, so we
+    // need to lock it in order to garuntee the state of tiff when
+    // it's being accessed by a thread.
+    tiff.synchronized { if (it.hasNext) Some(convertRaster(it.next)) else None }
   }
 
   def read(bounds: GridBounds[Long], bands: Seq[Int]): Option[Raster[MultibandTile]] = {
```

Comment:
The following test would fail: 

```scala
it("GeoTiffRasterSource readBounds() should be threadsafe") {
  val res: List[Future[Option[Raster[MultibandTile]]]] = (0 to 100).toList.map { _ => Future {
    source.read(source.gridBounds, 0 until source.bandCount)
  } }

  val fres: Future[List[Raster[MultibandTile]]] = Future.sequence(res).map(_.flatten)

  fres.map { rasters => assert(rasters.size == 101) }
}
```

Reply:
It is still a failing test: https://github.com/geotrellis/geotrellis-contrib/blob/d5c86d7664c58972538ea29bd34b5abba2363d8f/vlm/src/main/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSource.scala#L69